core/src/test/java/io/grpc/util/OutlierDetectionLoadBalancerTest.java

```diff
@@ -0,0 +1,978 @@
+        .setSuccessRateEjection(
+            new SuccessRateEjection.Builder()
+                .setMinimumHosts(3)
+                .setRequestVolume(20).build())
+        .setChildPolicy(new PolicySelection(roundRobinLbProvider, null)).build();
+
+    loadBalancer.handleResolvedAddresses(buildResolvedAddress(config, servers));
+
+    // We produce an outlier, but don't give it enough calls to reach the minimum volume.
+    generateLoad(
+        ImmutableMap.of(subchannel1, Status.DEADLINE_EXCEEDED),
+        ImmutableMap.of(subchannel1, 19));
+
+    // Move forward in time to a point where the detection timer has fired.
+    fakeClock.forwardTime(config.intervalSecs + 1, TimeUnit.SECONDS);
+
+    // The address should not have been ejected.
+    assertEjectedSubchannels(ImmutableSet.of());
+  }
+
+  /**
+   * The success rate algorithm does not apply if enough addresses have the required volume.
+   */
+  @Test
+  public void successRateOneOutlier_notEnoughAddressesWithVolume() {
+    OutlierDetectionLoadBalancerConfig config = new OutlierDetectionLoadBalancerConfig.Builder()
+        .setMaxEjectionPercent(50)
+        .setSuccessRateEjection(
+            new SuccessRateEjection.Builder()
+                .setMinimumHosts(6) // We don't have this many hosts...
+                .setRequestVolume(10).build())
+        .setChildPolicy(new PolicySelection(roundRobinLbProvider, null)).build();
+
+    loadBalancer.handleResolvedAddresses(buildResolvedAddress(config, servers));
+
+    generateLoad(ImmutableMap.of(subchannel1, Status.DEADLINE_EXCEEDED));
+
+    // Move forward in time to a point where the detection timer has fired.
+    fakeClock.forwardTime(config.intervalSecs + 1, TimeUnit.SECONDS);
+
+    // No subchannels should have been ejected.
+    assertEjectedSubchannels(ImmutableSet.of());
+  }
+
+  /**
+   * The enforcementPercentage configuration should be honored.
+   */
+  @Test
+  public void successRateOneOutlier_enforcementPercentage() {
+    OutlierDetectionLoadBalancerConfig config = new OutlierDetectionLoadBalancerConfig.Builder()
+        .setMaxEjectionPercent(50)
+        .setSuccessRateEjection(
+            new SuccessRateEjection.Builder()
+                .setMinimumHosts(3)
+                .setRequestVolume(10)
+                .setEnforcementPercentage(0)
+                .build())
+        .setChildPolicy(new PolicySelection(roundRobinLbProvider, null)).build();
+
+    loadBalancer.handleResolvedAddresses(buildResolvedAddress(config, servers));
+
+    generateLoad(ImmutableMap.of(subchannel1, Status.DEADLINE_EXCEEDED));
+
+    // Move forward in time to a point where the detection timer has fired.
+    fakeClock.forwardTime(config.intervalSecs + 1, TimeUnit.SECONDS);
+
+    // There is one outlier, but because enforcementPercentage is 0, nothing should be ejected.
+    assertEjectedSubchannels(ImmutableSet.of());
+  }
+
+  /**
+   * Two outliers get ejected.
+   */
+  @Test
+  public void successRateTwoOutliers() {
```

Comment:
Would be nice to have one more test with minimumHosts = 4 and 3 failing servers.  It should eject one and leave the other two.

Reply:
minimumHost refers to the minimum total amount of hosts that need to be present for the algorithm to consider any hosts, so in this scenario all three would get ejected (since we have five hosts).